homeassistant/components/recorder/core.py

```diff
@@ -1086,12 +1086,23 @@ def _process_state_changed_event_into_session(self, event: Event) -> None:
         entity_id = event.data["entity_id"]
 
         dbstate = States.from_event(event)
+        old_state = event.data["old_state"]
+
+        assert self.event_session is not None
+        session = self.event_session
 
         states_manager = self.states_manager
-        if old_state := states_manager.pop_pending(entity_id):
-            dbstate.old_state = old_state
+        if pending_state := states_manager.pop_pending(entity_id):
+            dbstate.old_state = pending_state
+            if old_state:
+                pending_state.last_reported_ts = old_state.last_reported_timestamp
         elif old_state_id := states_manager.pop_committed(entity_id):
             dbstate.old_state_id = old_state_id
+            if old_state:
+                session.query(States).filter_by(state_id=old_state_id).update(
+                    {States.last_reported_ts: old_state.last_reported_timestamp},
+                    synchronize_session=False,
+                )
```

Comment:
This generates a query like this:
```
2024-03-25 22:26:17.303 INFO     Recorder sqlalchemy.engine.Engine:base.py:1849 UPDATE states SET last_reported_ts=? WHERE states.state_id = ?
2024-03-25 22:26:17.303 INFO     Recorder sqlalchemy.engine.Engine:base.py:1849 [cached since 2s ago] (1711401976.303883, 6)
```